packages/core/src/indexing/sqlite/chronological-order.ts

```diff
@@ -0,0 +1,184 @@
+    } else {
+      return undefined
+    }
+  }
+}
+
+/**
+ * Prepare chronological cursor.
+ */
+function asChronologicalCursor(
+  input: { created_at: number; last_anchored_at: number } | undefined
+) {
+  if (!input) return undefined
+  return { created_at: input.created_at, last_anchored_at: input.last_anchored_at }
+}
+
+const CHRONOLOGICAL_ORDER = [
+  { column: 'last_anchored_at', order: 'DESC', nulls: 'last' }, // (`last_anchored_at` is null) DESC
+  { column: 'last_anchored_at', order: 'DESC' },
+  { column: 'created_at', order: 'DESC' },
+]
+
+/**
+ * Reverse ASC to DESC, and DESC to ASC in an order clause.
+ */
+function reverseOrder<T extends { order: string }>(entries: Array<T>): Array<T> {
+  const reverse = {
+    ASC: 'DESC',
+    DESC: 'ASC',
+  }
+  return entries.map((entry) => ({ ...entry, order: reverse[entry.order] }))
+}
+
+export class ChronologicalOrder {
+  constructor(private readonly dataSource: DataSource, private readonly knexConnection: Knex) {}
+
+  async page(query: BaseQuery & Pagination): Promise<Page<StreamID>> {
+    const pagination = parsePagination(query)
+    const paginationKind = pagination.kind
+    switch (paginationKind) {
+      case PaginationKind.FORWARD: {
+        const limit = pagination.first
+        const response = await this.query<Array<Selected>>(
+          this.chronologicalForwardQuery(query, pagination)
+        )
+        const entries = response.slice(0, limit)
+        const firstEntry = entries[0]
+        const lastEntry = entries[entries.length - 1]
+        return {
+          entries: entries.map((row) => StreamID.fromString(row.stream_id)),
+          pageInfo: {
+            hasNextPage: response.length > limit,
+            hasPreviousPage: false,
+            endCursor: Cursor.stringify(asChronologicalCursor(lastEntry)),
+            startCursor: Cursor.stringify(asChronologicalCursor(firstEntry)),
+          },
+        }
+      }
+      case PaginationKind.BACKWARD: {
+        const limit = pagination.last
+        const response = await this.query<Array<Selected>>(
+          this.chronologicalBackwardQuery(query, pagination)
+        )
+        const entries = response.slice(-limit)
+        const firstEntry = entries[0]
+        const lastEntry = entries[entries.length - 1]
+        return {
+          entries: entries.map((row) => StreamID.fromString(row.stream_id)),
+          pageInfo: {
+            hasNextPage: false,
+            hasPreviousPage: response.length > limit,
+            endCursor: Cursor.stringify(asChronologicalCursor(lastEntry)),
+            startCursor: Cursor.stringify(asChronologicalCursor(firstEntry)),
+          },
+        }
+      }
+      default:
+        throw new UnsupportedOrderingError(paginationKind)
+    }
+  }
+
+  private chronologicalForwardQuery(
+    query: BaseQuery,
+    pagination: ForwardPaginationQuery
+  ): Knex.QueryBuilder {
+    const tableName = asTableName(query.model)
+    const base = this.knexConnection
```

Comment:
I'm surprised you couldn't do this with TypeOrm.  TypeOrm does have a querybuilder API that can express some fairly complex queries.  See for example this query from the CAS: https://github.com/ceramicnetwork/ceramic-anchor-service/blob/f938eee1960591160126481fb580afe1706c1fbe/src/repositories/request-repository.ts#L122-L146

Reply:
The link shows that you could do a query if you create an entity class per table at development time. It is not our case here. TypeORM can not really create a query if you do not have a corresponding entity class. That's the reason to introduce here a dedicated query builder.

I am eagerly waiting for @pawartur and @willex spike results regarding this issue for the dedicated migration tool. If they can make TypeORM handle dynamic set of tables in dev-friendly way, knex is to be ditched. If not, this thing is here to stay.

@stephhuynh18 might have some additional input really soon about using TypeORM in CAS. As surprising, as it is, there is a chance, we might have to replace TypeORM with knex.

Reply:
Ah got it, that makes sense.